src/services/index.ts

```diff
@@ -0,0 +1,448 @@
+    }
+
+    // TODO: Send message on SQS  for real-time update
+    await dbUpdateLastSyncedEvent(mysql, fullNodeEvent.event.id);
+
+    await mysql.commit();
+  } catch (e) {
+    await mysql.rollback();
+    logger.error(e);
+
+    throw e;
+  } finally {
+    mysql.destroy();
+  }
+};
+
+export const handleVoidedTx = async (context: Context) => {
+  const mysql = await getDbConnection();
+  await mysql.beginTransaction();
+
+  try {
+    const fullNodeEvent = context.event as FullNodeEvent;
+
+    const {
+      hash,
+      outputs,
+      inputs,
+      tokens,
+    } = fullNodeEvent.event.data;
+
+    logger.debug(`Will handle voided tx for ${hash}`);
+
+    const dbTxOutputs: DbTxOutput[] = await getTxOutputsFromTx(mysql, hash);
+    const txOutputs: TxOutputWithIndex[] = prepareOutputs(outputs, tokens);
+    const txInputs: TxInput[] = prepareInputs(inputs, tokens);
+
+    // Set outputs as locked:
+
+    const txOutputsWithLocked = txOutputs.map((output) => {
+      const dbTxOutput = dbTxOutputs.find((_output) => _output.index === output.index);
+
+      if (!dbTxOutput) {
+        throw new Error('Transaction output different from database output!');
+      }
+
+      return {
+        ...output,
+        locked: dbTxOutput.locked,
+      };
+    });
+
+    const addressBalanceMap: StringMap<TokenBalanceMap> = getAddressBalanceMap(txInputs, txOutputsWithLocked);
+    await voidTransaction(mysql, hash, addressBalanceMap);
+    await markUtxosAsVoided(mysql, dbTxOutputs);
+
+    const addresses = Object.keys(addressBalanceMap);
+    await validateAddressBalances(mysql, addresses);
+
+    await dbUpdateLastSyncedEvent(mysql, fullNodeEvent.event.id);
+
+    logger.debug(`Voided tx ${hash}`);
+
+    await mysql.commit();
+  } catch (e) {
+    logger.debug(e);
+    await mysql.rollback();
+
+    throw e;
+  } finally {
+    mysql.destroy();
+  }
+};
+
+export const handleUnvoidedTx = async (context: Context) => {
+  const mysql = await getDbConnection();
+  await mysql.beginTransaction();
+
+  try {
+    const fullNodeEvent = context.event as FullNodeEvent;
+
+    const { hash } = fullNodeEvent.event.data;
+
+    logger.debug(`Tx ${hash} got unvoided, cleaning up the database.`);
+
+    await cleanupVoidedTx(mysql, hash);
+
+    logger.debug(`Unvoided tx ${hash}`);
+
+    await mysql.commit();
+  } catch (e) {
+    logger.debug(e);
+    await mysql.rollback();
+
+    throw e;
+  } finally {
+    mysql.destroy();
+  }
+};
+
+export const handleTxFirstBlock = async (context: Context) => {
+  const mysql = await getDbConnection();
+  await mysql.beginTransaction();
+
+  try {
+    const fullNodeEvent = context.event as FullNodeEvent;
+
+    const {
+      hash,
+      metadata,
+      timestamp,
+      version,
+      weight,
+    } = fullNodeEvent.event.data;
+
+    let height: number | null = metadata.height;
+
+    if (!metadata.first_block) {
+      height = null;
+    }
```

Comment:
Shouldn't we throw an error if `handleTxFirstBlock` is called and there's no `first_block`?

Reply:
Yes we should, done in 67b6f912198a5dd023acc704cbded8584393776b